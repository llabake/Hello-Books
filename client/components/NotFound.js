import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './common/SearchBar';

const NotFound = (props) =>
  <div className="container" style= {{ height: "95vh"}}>
  <div style={{marginLeft: "0px"}}>
    <h4>This is embarassing, something's missing.</h4>
    <b>404: page not found</b>
  </div>
    <div style={{margin: "10 55px 20 35px",}}>
      <SearchBar />
    </div>
    <div >
      <ul style={{"textAlign": "center"}}>
      <li style={{"display": "block"}}><Link to="/">Home</Link>
      </li>
      <li style={{"display": "block"}}><Link to="/allbooks">Books</Link>
      </li>
      <li style={{"display": "block"}}><Link to="/profile">Profile</Link>
      </li>
      </ul>
    </div>
  </div>

export default NotFound;

