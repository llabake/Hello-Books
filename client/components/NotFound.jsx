import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './common/SearchBar';

const NotFound = (props) =>
  <div className="container">
    <div style={{ marginLeft: "0px" }} id="not-found">
      <h4>This is embarassing, something's missing.</h4>
      <b>404: page not found</b>
    </div>
    <div>
      <SearchBar />
    </div>
    <div >
      <ul style={{ "textAlign": "center" }}>
        <li style={{ "display": "block" }}><Link to="/">Home</Link>
        </li>
        <li style={{ "display": "block" }}><Link to="/allbooks">Books</Link>
        </li>
        <li style={{ "display": "block" }}><Link to="/profile">Profile</Link>
        </li>
      </ul>
    </div>
  </div>

export default NotFound;

