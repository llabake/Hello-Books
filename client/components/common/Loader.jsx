import React from 'react';

import ajaxLoader from '../../media/ajax-loader.gif';

const Loader = () => {
  return (
    <div
      className="center-align"
      style={{ marginTop: '1em', marginBottom: '1em' }}
    >
      <img src={ajaxLoader} alt="Loading..." />
    </div>
  )
}

export default Loader;