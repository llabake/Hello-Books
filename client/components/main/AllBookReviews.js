import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'
import moment from 'moment';

import { deleteBookReview, loadReviewToEditAction } from '../../actions/bookAction';
import { getUser } from '../../helpers/utils';
import images from '../../media/images.jpg';
import EditReview from './EditReview';
import userImage from '../../media/user.png';

const captionStyle = {
  margin: '0',
  fontSize: '1.0625rem',
  fontStyle: 'normal',
  lineHeight: '32px',
  wordWrap: 'break-word',
  fontWeight: 700,
  maxWidth: '26rem',
  fontFamily: "'Lato', sans-serif",
}

const contentStyle = {
  fontSize: '.9375rem',
  lineHeight: '1.5',
  fontWeight: 400,
  fontStyle: 'normal',
  color: '21282d',
}


/**
 * 
 * 
 * @class AllBookReviews
 * @extends {Component}
 */
class AllBookReviews extends Component {
  /**
   * Creates an instance of AllBookReviews.
   * @param {any} props 
   * @memberof AllBookReviews
   */
  constructor(props) {
    super(props);

  }

  /**
   * @returns {void}
   * 
   * @param {any} reviewId 
   * @memberof AllBookReviews
   */
  handleDelete(reviewId) {
    this.props.deleteBookReview(reviewId)
  }

  /**
   * @returns {Object} review to be edited
   * 
   * @param {any} review 
   * @memberof AllBookReviews
   */
  handleLoadReviewToEdit(review) {
    this.props.loadReviewToEditAction(review)
  }

  /**
   * 
   * 
   * @returns {object} list of all review for a book
   * @memberof AllBookReviews
   */
  render() {
    const { user, reviews, editReview, reviewToEdit } = this.props
    return (
      <div>
        {editReview ? <EditReview review={reviewToEdit} /> : null}
        <div className="row">
          {reviews ?
            reviews.map((review, index) => {
              return <div className="col l10 offset-l1 m12 s12" key={index}>
                <div className='row'>
                  <div className='col s3' >
                    <div className='col s1' >
                      <img
                        className="responsive-img circle right"
                        src={review.user.image || userImage}
                        valign='top'
                        width="35"
                        height="35"
                        alt="reviewee image"
                      />
                    </div>
                    <div className='col s2' >
                      <span style={contentStyle}>
                        {review.user.username}
                      </span>
                      <br />
                      <span style={contentStyle}>
                        {/* <small style={{ borderLeft: '5px solid green' }}> 
                        {new Date(review.createdAt).toUTCString()}
                      </small> */}
                        <small style={{ borderLeft: '5px solid green' }}>
                          {/* {moment(review.createdAt).format()} */}
                          {/* {moment().startOf(review.createdAt).fromNow()} */}
                          {moment(review.createdAt, "YYYYMMDD").fromNow()}
                        </small>
                      </span>
                    </div>
                  </div>
                  <div className='col s9' >
                    <div>
                      <span style={captionStyle}>
                        {review.caption}
                      </span>
                      {user.id === review.user.id ?
                        <a className='right'>
                          <span className="write-review" onClick={this.handleEdit} >
                            <i
                              className="material-icons tiny pencil blue-text text-darken-2"
                              onClick={() => this.handleLoadReviewToEdit(review)}>
                              create</i>
                            <i
                              className="material-icons tiny pencil red-text"
                              onClick={() => this.handleDelete(review.id)}>
                              delete</i>
                          </span>

                        </a>
                        : null
                      }
                      <br />
                      <p style={contentStyle}>
                        {review.content}
                      </p>
                    </div>

                  </div>
                  {index ?
                    <div className='divider'
                      style={{ borderBottom: '1px solid #c8c8c8', overflow: 'visible' }}>
                    </div> : ''

                  }
                </div>

              </div>
            }) :
            null
          }
        </div>
      </div>
    )
  }

}

// const mapStateToProps = ({isLoading}) => ({
//   loading: isLoading
// });

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    allbookReview: state.bookReducer.allbookReview,
    loadTextArea: state.bookReducer.loadTextArea,
    loadAllReview: state.bookReducer.loadAllReview,
    editReview: state.bookReducer.editReview,
    reviewToEdit: state.bookReducer.reviewToEdit,
    user: state.userReducer.authUser
  };
}

const mapDispatchToProps = (dispatch) => {
  return {
    loadReviewToEditAction: (review) => dispatch(loadReviewToEditAction(review)),
    deleteBookReview: (reviewId) => dispatch(deleteBookReview(reviewId)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AllBookReviews);

