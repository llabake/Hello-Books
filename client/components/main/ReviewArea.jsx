import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'

import { reviewBook } from '../../actions/bookAction';
import TextAreaInput from '../common/TextAreaInput';
import TextInput from '../common/TextInput';
import InputValidator from '../../helpers/inputValidator';

const postReviewButton = {
  width: '25%',
  marginLeft: '36em'
}

/**
 * 
 * 
 * @class ReviewArea
 * @extends {Component}
 */
class ReviewArea extends Component {
  /**
   * Creates an instance of ReviewArea.
   * @param {any} props 
   * @memberof ReviewArea
   */
  constructor(props) {
    super (props);
    this.state = {
      errors: {},
      isValid: false,
      saving: false,
      content: '',
      caption: ''
    }

    this.handleAddReview = this.handleAddReview.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }
  


  /**
   * 
   * 
   * @returns {Object} response containing review validation 
   * @memberof ReviewArea
   */
  validate() {
    const { errors, isValid } = InputValidator.addReview(this.state);
    this.setState({ isValid, errors });
    return isValid;
  }

  /**
   * @returns {object} review object validity
   * 
   * @param {any} event 
   * @memberof ReviewArea
   */
  handleAddReview (event) {
    event.preventDefault();
    const reviewData = this.state
    this.props.reviewBook(this.props.bookId, reviewData)
    .then(() => this.clearForm());
  }

  /**
   * @returns {object} cleared form
   * 
   * @memberof ReviewArea
   */
  clearForm() {
    this.setState ({
      caption: '',
      content:'',
    })
  }

  /**
   * @returns {Object} review data
   * 
   * @param {any} event 
   * @memberof ReviewArea
   */
  handleChange (event) {
    this.setState({
      [event.target.name]: event.target.value
    }, () => this.validate());
  }
  /**
   * 
   * 
   * @returns {Object} review object
   * @memberof ReviewArea
   */
  render () {
    const { errors, isValid, saving } = this.state;
    return (
      <div>
        <form onSubmit= {this.handleAddReview}>
          <TextInput
          id = 'caption'
          type = 'text'
          icon = 'rate_review'
          name = 'caption'
          placeholder = 'Caption'
          onChange = {this.handleChange}
          value = {this.state.caption}
          errors = {errors.caption}
          />
          <TextAreaInput
          id='content'
          icon = 'rate_review'
          name = 'content'
          placeholder = 'Content'
          value = {this.state.content}
          onChange = {this.handleChange}
          errors = {errors.content}
          />
          <button type="submit" className="waves-effect waves-light btn right" 
            disabled= {!isValid || saving} style={postReviewButton}>Post Review
          </button>
        </form>
      </div>
    )
  }
  
}

const mapStateToProps = (state) => {
  return { errors: state.errors };
}

const mapDispatchToProps = (dispatch) => {
  return {
    reviewBook: (bookId, reviewData) => dispatch(reviewBook(bookId, reviewData))
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(ReviewArea);

