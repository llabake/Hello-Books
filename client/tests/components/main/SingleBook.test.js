import React from 'react';

import {
  SingleBook,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/main/SingleBook';
import NotFound from '../../../components/NotFound'
import Loader from '../../../components/common/Loader'
import ReviewArea from '../../../components/main/ReviewArea'
import AllBookReviews from '../../../components/main/AllBookReviews'

import adminData from '../../mocks/userData';
import userData from '../../mocks/userData';
import { bookAfterReviewAction } from '../../mocks/bookData'


function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    handleAddFavorite: jest.fn(),
    handleUpvote: jest.fn(),
    handleDownvote: jest.fn().mockImplementation(() => Promise.resolve(1)),
    handleShowAllReview: jest.fn(),
    handleBorrow: jest.fn(),
    handleShowAboutTheAuthor: jest.fn(),
    handleShowOverview: jest.fn(),
    fetchSingleBook: jest.fn(),
    favoriteABook: jest.fn(),
    upVoteBook: jest.fn(),
    downVoteBook: jest.fn().mockImplementation(() => Promise.resolve(1)),
    showReviewTextArea: jest.fn(),
    showAllReviews: jest.fn(),
    borrowBook: jest.fn(),
    errors: {},
    book: bookAfterReviewAction,
    match: {
      params: {
        bookId: 2
      }
    }
  };

  const state = {
    redirect: false,
    favorited: false,
    upvoted: false,
    downvoted: false,
    reviewed: false,
    showAboutTheAuthor: true,
    showOverview: false
  };
  const wrapper = shallow(<SingleBook {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('Single Book page', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should render a preloader', () => {
    const { wrapper } = setup();
    wrapper.setProps({ book: {}, loading: true });
    expect(wrapper.find(Loader).length).toBe(1)  
  });

  test('should render a load text area', () => {
    const { wrapper, props } = setup();
    wrapper.setProps({ loadTextArea: true });
    expect(wrapper.find(ReviewArea).length).toBe(1)

    // wrapper.instance().scrollIntoView();
    // const event = {
    //   preventDefault: jest.fn(),
    //   target: {}
    // }
    // const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    // wrapper.instance().handleShowReviewTextArea(event);
    // wrapper.instance.showReviewTextArea()
    // expect(dispatchSpy).toHaveBeenCalledTimes();
  });

  test('should render all book review load text area', () => {
    const { wrapper, props } = setup();
    wrapper.setProps({ loadAllReview: true });
    expect(wrapper.find(AllBookReviews).length).toBe(1)  
    // const event = {
    //   preventDefault: jest.fn(),
    //   target: {}
    // }
    // const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    // wrapper.instance().handleShowAllReview(event);
    // wrapper.instance.showAllReviews()
    // expect(dispatchSpy).toHaveBeenCalledTimes();
  });

  test('should render not found component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      loading: false,
      book: {},
      match: {
        params: {
          bookId: 22
        }
      },
      resourceNotFound: true,
      fetchSingleBook: jest.fn(() => Promise.resolve(1)),
    }

    const wrapper = shallow(<SingleBook {...props}  />);
    expect(wrapper.find(NotFound).length).toBe(1)

  })

  test('should call component did mount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });
    spy = jest.spyOn(SingleBook.prototype, 'componentDidMount')
    const { wrapper, props } = setup();
    expect(wrapper).toBeDefined();
    expect(SingleBook.prototype.componentDidMount).toHaveBeenCalledTimes(1)
  })
  
  // test('should dispatch an action to upvote for book when the method is called', () => {
  //   const { wrapper, props, state  } = setup();
  //   const event = {
  //     preventDefault: jest.fn(),
  //     target: {}
  //   }
  //   const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
  //   wrapper.instance().handleUpvote(event);
  //   wrapper.instance.upVoteBook(2)
  //   // expect(dispatchSpy).toHaveBeenCalledTimes();

  //   // props.upVoteBook.mockReset();
  //   // wrapper.instance().handleUpvote(event)
  //   // expect(props.upVoteBook).toHaveBeenCalled()
  // });

  // test('should dispatch an action to downvote for book when the method is called', () => {
  //   const { wrapper, props } = setup();
  //   const event = {
  //     preventDefault: jest.fn(),
  //     target: {}
  //   }
  //   // const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
  //   // wrapper.instance().handleDownvote(event);
  //   // wrapper.instance.downVoteBook(2)
  //   // expect(dispatchSpy).toHaveBeenCalledTimes();

  //   props.downVoteBook.mockReset();
  //   wrapper.instance().handleDownvote(event)
  //   expect(props.downVoteBook).toHaveBeenCalled()
  // });


  test('should dispatch an action to favorite a book when the method is called', () => {
    const { wrapper , props } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {}
    }
    props.favoriteABook.mockReset();
    wrapper.instance().handleAddFavorite(event)
    expect(props.favoriteABook).toHaveBeenCalled()
  });


  // test('should dispatch an action to borrow a book when the method is called', () => {
  //   const { wrapper } = setup();

  //   const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
  //   wrapper.instance().handleBorrow();
  //   wrapper.instance.borrowBook(2)
  //   // expect(dispatchSpy).toHaveBeenCalledTimes();


  // //   const handleBorrow =
  // //   jest.spyOn(wrapper.instance(), 'handleBorrow');
  // //   handleBorrow();
  // // wrapper.find('.btn').simulate('click');
  // // expect(wrapper.instance().handleBorrow).toHaveBeenCalled();
  // });

  test('should display about the author tab', () => {
    const { wrapper, props, state } = setup();
    wrapper.setState({
      showAboutTheAuthor: true,
      showOverview: false
    });
    props.handleShowAboutTheAuthor.mockReset();
    wrapper.instance().handleShowAboutTheAuthor()
    expect(wrapper.instance().state.showAboutTheAuthor).toBe(true);
    expect(wrapper.instance().state.showOverview).toBe(false);
  })

  test('should display overview tab', () => {
    const { wrapper, props, state } = setup();
    wrapper.setState({
      showAboutTheAuthor: false,
      showOverview: true
    });
    props.handleShowOverview.mockReset();
    wrapper.instance().handleShowOverview()
    expect(wrapper.instance().state.showAboutTheAuthor).toBe(false);
    expect(wrapper.instance().state.showOverview).toBe(true);
  })

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchSingleBook');

    const {
      fetchSingleBook,
      favoriteABook,
      upVoteBook,
      downVoteBook,
      showReviewTextArea,
      showAllReviews,
      borrowBook
    } = mapDispatchToProps(dispatch);
    fetchSingleBook();
    favoriteABook();
    upVoteBook();
    downVoteBook();
    showReviewTextArea();
    showAllReviews();
    borrowBook();
    expect(dispatch).toHaveBeenCalled();
  });

  test('mapStateToProps', () => {


    const storeState = {
      userReducer: {
        authUser: {},
        authenticated: false
      },
      bookReducer: {
        book: {}
      },
    };
  
    expect(mapStateToProps(storeState).user).toEqual({});
    expect(mapStateToProps(storeState).book).toEqual({});
    // expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('book');
    // expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('loading');
    // expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('loadTextArea');
    // expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('loadAllReview');
    // expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('resourceNotFound');
    // expect(mapStateToProps({ userReducer: {} })).toHaveProperty('user');
    // expect(mapStateToProps({ userReducer: {} })).toHaveProperty('authenticated');
  });




})