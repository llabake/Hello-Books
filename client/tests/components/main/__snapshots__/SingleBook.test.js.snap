// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Book page should render and not crash 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="divider"
    />
    <div
      className="book-detail center-align"
    >
      <h1
        id="title"
      >
        dont be sad 
      </h1>
      <div>
        <span
          className="detail"
          id="author-top"
        >
          by 
          <span
            className="author-name"
          >
            Aaidh ibn Abdullah al-Qarni
          </span>
        </span>
      </div>
      <div
        className="divider"
      />
    </div>
    <div
      className="section "
    >
      <div
        className="book"
      >
        <div
          className="book-container"
        >
          <div
            className="row "
          >
            <div
              className="col l5 show-border"
            >
              <img
                className="materialboxed"
                src="https://res.cloudinary.com/sardaunan/image/upload/v1520866315/hbg1rryhpzabtnh2akzx.webpk"
                width="250"
              />
            </div>
            <div
              className="col  l7"
            >
              <div
                className="section"
              >
                <div
                  className="section1"
                >
                  <div>
                    <a
                      className="write-review "
                      onClick={[Function]}
                      style={
                        Object {
                          "cursor": "pointer",
                        }
                      }
                    >
                      <i
                        className=" material-icons pencil medium"
                      >
                        favorite
                      </i>
                      Add to Favorite
                    </a>
                  </div>
                  <form
                    id="form-lend"
                  >
                    <a
                      className="waves-effect waves-light btn"
                      onClick={[Function]}
                    >
                      Borrow
                    </a>
                  </form>
                </div>
              </div>
              <div
                className="divider"
              />
              <div
                className="section"
              >
                <div
                  className="section2"
                >
                  <span>
                     Quantity In Stock: 
                  </span>
                   
                  <span>
                    4
                  </span>
                  <span>
                     Status: 
                  </span>
                  <span>
                    Available
                  </span>
                </div>
              </div>
              <div
                className="divider"
              />
              <div
                className="section"
              >
                <div
                  className="section3"
                >
                  <span>
                    dont be sad 
                     has been borrowed 
                  </span>
                  <span>
                    0
                  </span>
                   times 
                  <span>
                    <a
                      className="write-review"
                      onClick={[Function]}
                    >
                      <i
                        className=" material-icons tiny pencil"
                      >
                        create
                      </i>
                      Review?
                    </a>
                  </span>
                  <br />
                  <div
                    style={
                      Object {
                        "cursor": "pointer",
                      }
                    }
                  >
                    <a
                      className="upvote"
                      onClick={[Function]}
                    >
                      <i
                        className=" material-icons pencil small"
                      >
                        thumb_up
                      </i>
                      1
                    </a>
                    <a
                      className=" upvote"
                      onClick={[Function]}
                    >
                      <i
                        className="material-icons pencil small"
                      >
                        thumb_down
                      </i>
                      0
                    </a>
                    <a
                      className=" upvote"
                      onClick={[Function]}
                    >
                      <i
                        className=" material-icons pencil small"
                      >
                        comment
                      </i>
                      1
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="divider"
    />
    <div
      className="section tab"
    >
      <div
        className="row"
      >
        <div
          className="col s12"
        >
          <ul
            className="tabs .tabs-fixed-width"
            id="tab"
          >
            <li
              className="tab col s6"
            >
              <a
                className="active"
                onClick={[Function]}
              >
                About the author
              </a>
            </li>
            <li
              className="tab col s6"
            >
              <a
                onClick={[Function]}
              >
                Overview
              </a>
            </li>
          </ul>
        </div>
        <div
          id="about"
        >
          <div
            className="col s12"
          >
            <div
              className="row "
            >
              <div>
                <h5
                  className="center-align"
                >
                  About the Author
                </h5>
              </div>
            </div>
            <span>
              alrefi is a god writer ten 
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      id="review-area"
    >
      <h3
        id="review"
      >
         Reviews 
      </h3>
      <a
        className="write-review"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        View all
      </a>
      <a
        className="write-review"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <i
          className=" material-icons tiny pencil"
        >
          create
        </i>
        Write your review
      </a>
      <div
        className="divider"
      />
      <div
        className="section "
      >
        <div
          className="row"
        >
          <div
            className="col s4"
            key="0"
          >
            <h4
              className="rev-title"
            >
              “
              escape
              ”
            </h4>
             
            motivation and commitment
          </div>
        </div>
      </div>
    </div>
    <div
      className="divider"
    />
    <div
      className="section similar"
    >
      <h3>
        You may also be interested in...
      </h3>
      <Connect(Carousel) />
    </div>
  </div>
</div>
`;
